### ISSUE TITLE
Extend rule: max-params

### ISSUE THREAD

**Issue by zimme** [eslint/eslint#9562]
**What rule do you want to change?**
max-params

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
I would hope to see a new option `ignore-callbacks`

**Please provide some example code that this change will affect:**

```js
// max-params, warn, 1

function foo(bar, baz) { ... }

function foo({ bar, baz }) { ... }

getData((error, result) => { ... });
```

**What does the rule currently do for this code?**
First function gets a warning because it has more than 1 argument.
Second function is ok as it only has one argument.
Callback gets a warning as it has 2 arguments.

**What will the rule do after it's changed?**
First function gets a warning.
Second function is ok
Callback is ok

I usually have `max-params` set to `warn` with `1` as I like to force the use of "named" params via objects.

It would be nice to have an option (`ignore-callbacks`?) for `max-params` to ignore functions that are used as callbacks.
I'm not in control of the signature of expected callback functions and I keep getting a lot of warnings for these functions.

**Comment by zimme** [comment]
If this is of interest I could try a PR for this with some help.

**Comment by platinumazure** [comment]
Hi @zimme, thanks for the issue!

Would you mind editing your original post to conform to our [rule change proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md)? It makes evaluation of proposed rule changes so much easier for us :smile:

Personally, I could see us adding an exceptions string array or regex which would not report functions with certain names. Or, I could also see some sort of option representing number and/or names of parameters which could also be used to avoid reporting certain cases... Do think on it a bit and see if you can come up with a few possibilities when you fill in the rule change proposal template :smile:

**Comment by zimme** [comment]
There, I've updated my original post. Let's discuss. =)

**Comment by ilyavolodin** [comment]
I don't think this is possible to implement correctly. It would only work if you are using it as callback with arrow function, but not as a declared function:

```js
function foo(bar, baz) { ... };
getData(foo);
```

I'm not sure it would be a good idea to add an option that wouldn't work for every case. I know arrow functions are becoming more popular, especially for callbacks, but there are still plenty of cases where you would use named function as callback parameter.

**Comment by platinumazure** [comment]
@zimme I think this might be best implemented as a custom rule. You could use [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer) to extend the max-params rule and filter out reports that point to callback functions.

Another possibility would be to use [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) to flag functions with different parameter counts based on other expressions you specify in the AST selector syntax. For example, you could go by parameter names instead of whether or not the function is a callback.

**Comment by zimme** [comment]
Thanks for the feedback, I'll try and make a custom rule for this and see where I end up :p

**Comment by ilyavolodin** [comment]
Great! If you end up creating a custom rule for this, can you post a link here, so that other might learn about it. Closing this issue, as it seems resolved.
